lib/mongodb/queries.nit

```diff
@@ -0,0 +1,437 @@
+#
+# Using the `queries` framework we can get from this:
+#
+# ~~~nitish
+# var exists = new JsonObject
+# exists["$exists"] = true
+#
+# var query = new JsonObject
+# query["login"] = "Morriar"
+# query["email"] = exists
+#
+# collection.find(query)
+# ~~~
+#
+# To this:
+#
+# ~~~nitish
+# collection.find((new MongoMatch).eq("login", "Morriar").exists("email", true))
+# ~~~
+#
+# The framework provides three classes:
+# * `MongoMatch` the base query that can be used with most Mongo services
+# * `MongoPipeline` the array of queries that is expected by `MongoCollection::aggregate`
+# * `MongoGroup` the group query for a `MongoPipeline`
+module queries
+
+import mongodb
+
+# A basic match query
+#
+# `MongoMatch` is used with most of the Mongo services like `find`, `find_all`,
+# `remove` etc.
+#
+# Building a query can be done with the fluent interface:
+#
+# ~~~
+# var query = (new MongoMatch).
+#	eq("login", "Morriar").
+#	gt("age", 18).
+#	exists("email", true).
+#	is_in("status", [1, 2, 3, 4])
+# ~~~
+class MongoMatch
+	super JsonObject
+
+	private fun op(name: String, field: String, value: nullable Jsonable): MongoMatch do
+		var q = new JsonObject
+		q["${name}"] = value
+		self[field] = q
+		return self
+	end
+
+	# Match documents where `field` equals `value`
+	#
+	# ~~~json
+	# {field: {$eq: value} }
+	# ~~~
+	fun eq(field: String, value: nullable Jsonable): MongoMatch do
+		self[field] = value
+		return self
+	end
+
+	# Match documents where `field` not equals `value`
+	#
+	# ~~~json
+	# {field: {$ne: value} }
+	# ~~~
+	fun ne(field: String, value: nullable Jsonable): MongoMatch do
+		op("ne", field, value)
+		return self
+	end
+
+	# Match documents where `field` is greater than `value`
+	#
+	# ~~~json
+	# {field: {$gt: value} }
+	# ~~~
+	fun gt(field: String, value: nullable Jsonable): MongoMatch do
+		op("gt", field, value)
+		return self
+	end
+
+	# Match documents where `field` is greater or equal to `value`
+	#
+	# ~~~json
+	# {field: {$gte: value} }
+	# ~~~
+	fun gte(field: String, value: nullable Jsonable): MongoMatch do
+		op("gte", field, value)
+		return self
+	end
+
+	# Match documents where `field` is less than `value`
+	#
+	# ~~~json
+	# {field: {$lt: value} }
+	# ~~~
+	fun lt(field: String, value: nullable Jsonable): MongoMatch do
+		op("lt", field, value)
+		return self
+	end
+
+	# Match documents where `field` is less or equal to `value`
+	#
+	# ~~~json
+	# {field: {$lte: value} }
+	# ~~~
+	fun lte(field: String, value: nullable Jsonable): MongoMatch do
+		op("lte", field, value)
+		return self
+	end
+
+	# Match documents where `field` exists or not
+	#
+	# ~~~json
+	# {field: {$exists: value} }
+	# ~~~
+	#
+	# When `exists` is true, `$exists` matches the documents that contain the
+	# field, including documents where the field value is null.
+	# If <boolean> is false, the query returns only the documents that do not
+	# contain the field.
+	fun exists(field: String, value: nullable Jsonable): MongoMatch do
```

Comment:
I have an hard time with this one. Must `value` be either `true` or `false`? If yes, why not type with `Bool`.
And why not drop the `value` completely and add another method `not_exists`?